### PR TITLE
Dataset Copy API

### PR DESCRIPTION
**Pull request by sakshibobade21**
## Proposed changes

This PR addresses Issue: [*Link to Github issue within https://github.com/zowe/zss/issues* if any]

This PR depends upon the following PRs:

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Change in a documentation
- [ ] Refactor the code 
- [ ] Chore, repository cleanup, updates the dependencies.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## PR Checklist
Please delete options that are not relevant.
- [ ] If the changes in this PR are meant for the next release / mainline, this PR targets the "staging" branch.
- [ ] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zss/blob/v2.x/master/CONTRIBUTING.md))
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] video or image is included if visual changes are made
- [ ] Relevant update to CHANGELOG.md
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works, or describe a test method below

### CODE REVIEW

File: c/datasetService.c [hunk 1]
```diff
@@ -56,7 +56,7 @@ static int serveDatasetMetadata(HttpService *service, HttpResponse *response) {
     char *l1 = stringListPrint(request->parsedFile, 1, 1, "/", 0); //expect name or hlq
     zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_DEBUG, "L1=%s\n", l1);
     if (!strcmp(l1, "name")){
-      respondWithDatasetMetadata(response);
+      getDatasetMetadataFromRequest(response);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: get implies something is returned to the function, but it isnt?
so we should still say something like, respondWithDatasetMadataFromRequest(response);
Or just don't change the name?

File: c/datasetService.c [hunk 2]
```diff
@@ -124,11 +124,10 @@ static int serveDatasetContents(HttpService *service, HttpResponse *response){
     char *filename = stringConcatenate(response->slh, filenamep1, "'");
     zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_DEBUG, "Allocating dataset: %s\n", filename);
     fflush(stdout);
-    newDataset(response, filename, TRUE);
+    newDatasetFromRequest(response, filename, TRUE);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: maybe this should be... createDatasetAndRespond()

File: c/datasetService.c [hunk 3]
```diff
@@ -146,6 +145,40 @@ static int serveDatasetContents(HttpService *service, HttpResponse *response){
...
+    fflush(stdout);
+    copyDataset(response, datasetName, newDatasetName);
+  } else {
+    jsonPrinter *out = respondWithJsonPrinter(response);
+
+    setContentType(response, "text/json");
+    setResponseStatus(response, 405, "Method Not Allowed");
+    addStringHeader(response, "Server", "jdmfws");
+    addStringHeader(response, "Transfer-Encoding", "chunked");
+    addStringHeader(response, "Allow", "GET, DELETE, POST");
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: should just be "POST", due to above

File: c/datasetService.c [hunk 4]
```diff
@@ -146,6 +145,40 @@ static int serveDatasetContents(HttpService *service, HttpResponse *response){
...
+    writeHeader(response);
+
+    jsonStart(out);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: sending `{}` may confuse people. either do a blank body (no json printer at all) or print an actual error.

File: c/datasetService.c [hunk 5]
```diff
@@ -146,6 +145,40 @@ static int serveDatasetContents(HttpService *service, HttpResponse *response){
...
+
+  if (!strcmp(request->method, methodPOST)){
+    char *l1 = stringListPrint(request->parsedFile, 1, 1, "/", 0);
+    char *percentDecoded = cleanURLParamValue(response->slh, l1);
+    char *datasetNameP1 = stringConcatenate(response->slh, "//'", percentDecoded);
+    char *datasetName = stringConcatenate(response->slh, datasetNameP1, "'");
+    char *newDataset = getQueryParam(response->request, "newDataset");
+    char *newDatasetNameP1 = stringConcatenate(response->slh, "//'", newDataset);
+    char *newDatasetName = stringConcatenate(response->slh, newDatasetNameP1, "'");
+    fflush(stdout);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: the use of `fflush(stdout)` is if you were doing a lot of `zowelog` or `printf` and didn't use `fflush`, then the messages wouldnt be printed to the screen until the OS decides to, because stdout could be buffered.
`fflush` forces the buffer to flush, so messages get printed out immediately.
Point is, you may not need this here.

File: c/datasetjson.c [hunk 6]
```diff
@@ -900,11 +904,11 @@ static void respondWithDYNALLOCError(HttpResponse *response,
 #define IS_DAMEMBER_EMPTY($member) \
   (!memcmp(&($member), &(DynallocMemberName){"        "}, sizeof($member)))
 
-static void updateDatasetWithJSONInternal(HttpResponse* response,
+static int updateDatasetWithJSONInternal(HttpResponse* response,
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: it looks like the return `int` is only ever `-1`, so you may want to add some different ints for different error reasons, like `-2`, etc... or more commonly, you could do:

0=ok
1=problem
2= more problems
3....
Or,

0=ok
4=warn
8=severe

Comment (1000TurquoisePogs) [reply c/datasetjson.c]: possibly, you could change the function arguments from having `recordsUpdated` to returning it as an int? so, if it's 0 or negative, then you know it failed.

Comment (1000TurquoisePogs) [reply c/datasetjson.c]: this function needs a code comment that says:

if successful, no http content is written.
if unsuccessful, respondWithError is used, so the caller should not ALSO respond.

File: c/datasetjson.c [hunk 7]
```diff
@@ -1451,6 +1449,24 @@ void updateDataset(HttpResponse* response, char* absolutePath, int jsonMode) {
     respondWithError(response,HTTP_STATUS_INTERNAL_SERVER_ERROR,"Could not translate character set to EBCDIC");    
   }
   safeFree(convertedBody,conversionBufferLength);
+
+  if(rc != -1) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: if changing the code to have more/different rcs, its often the case we just do `if (rc != 0)`

File: c/datasetjson.c [hunk 8]
```diff
@@ -2220,6 +2234,464 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+    respondWithError(response,HTTP_STATUS_BAD_REQUEST,"No target dataset name given");
+    return;
+  }
+
+  if(!isDatasetPathValid(sourceDataset)){
+    respondWithError(response,HTTP_STATUS_BAD_REQUEST,"Invalid dataset path");
+    return;
+  }
+
+  if(!isDatasetPathValid(targetDataset)){
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: then, if valid, check if it exists, and if it exists, respond with error (cant copy because already exists)

File: c/datasetjson.c [hunk 9]
```diff
@@ -1083,7 +1087,7 @@ static void updateDatasetWithJSONInternal(HttpResponse* response,
       zowelog(NULL, LOG_COMP_RESTDATASET, ZOWE_LOG_DEBUG, "Error writing to dataset, rc=%d\n", bytesWritten);
       respondWithError(response,HTTP_STATUS_INTERNAL_SERVER_ERROR,"Error writing to dataset");
       fclose(outDataset);
-      break;
+      return -1;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: this loop should be extracted into a function that doesnt need json, or something like that... because this loop is the way to stream a write without needing an entire buffer of an entire dataset, so you'll want to reuse this type of logic for the copy.

File: c/datasetjson.c [hunk 10]
```diff
@@ -2220,6 +2234,464 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+  rc = setAttributesForDatasetCopy(response, datasetAttrBuffer, datasetAttributes);
+
+  if(rc == -1) {
+    return;
+  }
+
+  // To read the source dataset content in the buffer
+  JsonBuffer *datasetContentBuffer = makeJsonBuffer();
+  jsonPrinter *p = makeBufferNativeJsonPrinter(CCSID_UTF_8, datasetContentBuffer);
+  rc = readDatasetContent(response, sourceDataset, p);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: this is the most important part of the PR, and needs to be fixed.
you are reading the dataset into a buffer. doesnt matter if its json or text or anything else, the problem with making a buffer the size of the dataset, is this puts the entire contents of the dataset into RAM, and datasets could be gigabytes, or terabytes and you'll run out of ram and crash.


so you need to stream instead of buffer it all.
basically,

int bytesRead, bytesWritten = 4096;
char buffer[4096];
while (bytesRead == 4096 && bytesWritten == 4096) {
  bytesRead = read(dataset, &buffer);
  bytesWritten = write(newDataset, &buffer);
}

**Review by 1000TurquoisePogs — COMMENTED** [review]
we'll re-review once readDatasetContent and pasteDatasetContent are rewritten to be streaming instead of buffer-entire-dataset code.

File: c/datasetService.c [hunk 11]
```diff
@@ -146,6 +144,33 @@ static int serveDatasetContents(HttpService *service, HttpResponse *response){
...
+
+  if (!strcmp(request->method, methodPOST)){
+    char *l1 = stringListPrint(request->parsedFile, 1, 1, "/", 0);
+    char *percentDecoded = cleanURLParamValue(response->slh, l1);
+    char *datasetNameP1 = stringConcatenate(response->slh, "//'", percentDecoded);
+    char *datasetName = stringConcatenate(response->slh, datasetNameP1, "'");
+    char *newDataset = getQueryParam(response->request, "newDataset");
+    char *newDatasetNameP1 = stringConcatenate(response->slh, "//'", newDataset);
+    char *newDatasetName = stringConcatenate(response->slh, newDatasetNameP1, "'");
+    copyDataset(response, datasetName, newDatasetName);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c]: copyDatasetAndRespond

File: c/datasetService.c [hunk 12]
```diff
@@ -204,6 +229,19 @@ void installDatasetContentsService(HttpServer *server) {
...
+void installDatasetCopyService(HttpServer *server) {
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_INFO, ZSS_LOG_INSTALL_MSG, "dataset copy paste");
+  zowelog(NULL, LOG_COMP_ID_MVD_SERVER, ZOWE_LOG_DEBUG2, "begin %s\n", __FUNCTION__);
+
+  HttpService *httpService = makeGeneratedService("datasetCopy", "/datasetCopy/**");
+  httpService->authType = SERVICE_AUTH_NATIVE_WITH_SESSION_TOKEN;
+  httpService->runInSubtask = TRUE;
+  httpService->doImpersonation = TRUE;
+  httpService->serviceFunction = serveDatasetCopy;
+  httpService->paramSpecList = makeStringParamSpec("force",SERVICE_ARG_OPTIONAL, NULL);
```
Comment (1000TurquoisePogs) [inline c/datasetService.c:240]: comment out?
or, try this:

if (!force && targetDatasetExists) {
  respondWithError(cannot copy, target already exists)
} else {
  doCopy
}

File: c/datasetjson.c [hunk 13]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+  }
+
+  if(!strcmp(organization, "sequential")) {
+    organization = "PS";
+    dsnt = "";
+  } else {
+    organization = "PO";
+    dsnt = (isPDSE) ? "PDSE" : "PDS";
+    dirBlock = 10;
+    respondWithError(response, HTTP_STATUS_INTERNAL_SERVER_ERROR, "Do not support copy-paste for PDS(E) Dataset");
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: might want to put a TODO here so you remember that it needs to be implemented later.

File: c/datasetjson.c [hunk 14]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  finishResponse(response);
+
+  fclose(inDataset);
+  fclose(outDataset);
+
+  return;
+}
+
+void readAndWriteToDataset(HttpResponse *response, char* sourceDataset, char* targetDataset) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: +respond (in case someone thought this was an internal function)

File: c/datasetjson.c [hunk 15]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  char ddPath[16];
+  snprintf(ddPath, sizeof(ddPath), "DD:%8.8s", ddName.value);
+
+  int lrecl = getLreclOrRespondError(response, &dsn, ddPath);
+  if (!lrecl) {
+    return;
+  }
+
+  readAndWriteToDatasetInternal(response, ddPath, lrecl, targetDataset);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: this also responds. should it? maybe not if it's internal? dont have a strong opinion, as long as you name your functions well.

File: c/datasetjson.c [hunk 16]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  DDName ddName;
+  memcpy(&ddName.value, &daDDname.name, sizeof(ddName.value));
+
+  char ddPath[16];
+  snprintf(ddPath, sizeof(ddPath), "DD:%8.8s", ddName.value);
+
+  int lrecl = getLreclOrRespondError(response, &dsn, ddPath);
+  if (!lrecl) {
+    return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: i think a `dynallocUnallocDatasetByDDName` needs to occur with this return, else the server holds some sort of lock on the dataset?

we see a `dynallocUnallocDatasetByDDName` on line 2448 so just make sure there's one here too, i think.

File: c/datasetjson.c [hunk 17]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  if (json) {
+    if (jsonIsObject(json)){
+      JsonObject *jsonObject = jsonAsObject(json);
+      //Get array of datasets
+      JsonArray *datasetArray = jsonObjectGetArray(jsonObject,"datasets");
+      datasetCount = jsonArrayGetCount(datasetArray);
+    }
+  }
+  SLHFree(slh);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2746]: also do a `safeFree((char *)datasetAttrBuffer, datasetAttrBuffer->size)`, because we saw that `makeJsonBuffer` does a `safeMalloc` to get you that `JsonBuffer`

File: c/datasetjson.c [hunk 18]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
-  int memberNameLength = 0;
+  extractDatasetAndMemberName(sourceDataset, &dsnName, &memName);
 
-  extractDatasetAndMemberName(absDsPath, &dsnName, &memName);
-  memberNameLength = (unsigned int)rParenIndex  - (unsigned int)lParenIndex -1;
-  
-  HttpRequestParam *addQualifiersParam = getCheckedParam(request,"addQualifiers");
-  char *addQualifiersArg = (addQualifiersParam ? addQualifiersParam->stringValue : NULL);
+  // Buffer to save attributes for source dataset
+  JsonBuffer *datasetAttrBuffer = makeJsonBuffer();
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2940]: it looks like this needs to be `safeFree`.

File: c/datasetjson.c [hunk 19]
```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+        JsonObject *recfm = jsonObjectGetObject(jsonDatasetObject,"recfm");
+        *recordLength = jsonObjectGetString(recfm,"recordLength");
+        *isBlocked = jsonObjectGetBoolean(recfm,"isBlocked");
+      }
+    }
+  }
+  SLHFree(slh);
+}
+
+int setAttributesForDatasetCopy(HttpResponse *response, JsonBuffer *buffer, char* datasetAttributes) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: can respond, make note of that.

File: c/datasetService.c (hunk 12 above)
Comment (sakshibobade21) [reply c/datasetService.c:240]: I actually thought we will need `force` for the etag. But we do not. Because the new dataset will be created and we don't need to compare the etag.
And we have already checked if the target dataset exists or not in  `checkIfDatasetExists` function.

**Review by 1000TurquoisePogs — APPROVED** [review]
This looks good to me now, but I'm still getting better at doing such review.
I'm not going to merge this yet until we see it live in dependent plugins (which is in progress)
If someone else wants to take a look meanwhile, that's nice too, otherwise I'll merge this when I see it working as intended.

File: c/datasetjson.c [hunk 20]
```diff
@@ -2312,7 +2314,50 @@ int setAttrForDSCopyAndRespondIfError(HttpResponse *response, JsonBuffer *buffer
...
+      //Get array of datasets
+      JsonArray *datasetArray = jsonObjectGetArray(jsonObject,"datasets");
+      int i = 0;
+      Json *element = jsonArrayGetItem(datasetArray,i);
+      if(element && jsonIsObject(element)) {
+        JsonObject *jsonDatasetObject = jsonAsObject(element);
+        // Get dsorg object
+        JsonObject *dsOrg = jsonObjectGetObject(jsonDatasetObject,"dsorg");
+        targetRecLen = jsonObjectGetNumber(dsOrg,"maxRecordLen");
+        printf("-----TARGET REC LENGTH : %d", targetRecLen);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: printfs that you dont need could be turned into zowelog( ... _DEBUG) statements.

File: c/datasetjson.c [hunk 21]
```diff
@@ -2336,28 +2382,46 @@ void streamDatasetForCopyAndRespond(HttpResponse *response, char *sourceDataset,
...
 
   while (!feof(inDataset)){
-    bytesRead = fread(buffer,1,recordLength,inDataset);
+    bytesRead = fread(buffer,1,sourceRecordLen,inDataset);
+    printf("---BYTES READ: %d\n", bytesRead);
     if (bytesRead > 0 && !ferror(inDataset)) {
       bytesWritten = fwrite(buffer,1,bytesRead,outDataset);
+      printf("---BYTES WRITTEN: %d\n", bytesWritten);
+      if(ferror(outDataset)) {
+        printf("---ferror outDataset\n");
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: more printfs to do something about

File: c/datasetjson.c [hunk 22]
```diff
@@ -908,6 +916,44 @@ static void respondWithDYNALLOCError(HttpResponse *response,
...
 }
 
+static void getDYNALLOCErrorCodeAndMsg(int rc, int sysRC, int sysRSN,
+                                       const DynallocDatasetName *dsn,
+                                       const DynallocMemberName *member,
+                                       const char *site,
+                                       char* responseMessage,
+                                       int* responseCode) {
+
+  if (rc ==  RC_DYNALLOC_SVC99_FAILED && sysRC == 4) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:1002]: this is code duplication with above, let's try to put this into a common function, because as we discover more reasons to print out, we'll want to add them in both places and it could be a maintenance nightmare later.

Comment (1000TurquoisePogs) [reply c/datasetjson.c:1002]: perhaps this is just the difference of whether to printf or to respond with message, depending on which object got passed in.

Comment (1000TurquoisePogs) [reply c/datasetjson.c:1002]: get rid of 
static void respondWithDYNALLOCError(HttpResponse *response,
                                     int rc, int sysRC, int sysRSN,
                                     const DynallocDatasetName *dsn,
                                     const DynallocMemberName *member,
                                     const char *site) {

just have 

static void getDYNALLOCErrorCodeAndMsg(int rc, int sysRC, int sysRSN,
                                       const DynallocDatasetName *dsn,
                                       const DynallocMemberName *member,
                                       const char *site,
                                       char* responseMessage,
                                       int* responseCode) {

then, users of respondWithDYNALLOCError could just receive `responseMessage` and `responseCode` and do `respondWithMessage` by themselves.

File: c/datasetjson.c [hunk 23]
```diff
@@ -2336,30 +2449,47 @@ void streamDatasetForCopyAndRespond(HttpResponse *response, char *sourceDataset,
...
+
+  int bufferSize = sourceRecordLen+1;
   char buffer[bufferSize];
   int bytesRead = 0;
   int bytesWritten = 0;
   int recordsWritten = 0;
 
   while (!feof(inDataset)){
-    bytesRead = fread(buffer,1,recordLength,inDataset);
+    bytesRead = fread(buffer,1,sourceRecordLen,inDataset);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2565]: when source record length is less than target, on each line it would be best to right-pad with 0x40 (space) during the `fwrite`

File: c/datasetjson.c [hunk 24]
```diff
@@ -2501,9 +2643,120 @@ bool checkIfDatasetExists(char* dataset) {
...
+    setDefaultJSONRESTHeaders(response);
+    writeHeader(response);
+    jsonStart(p);
+
+    jsonAddString(p, "msg", msgBuffer);
+    jsonAddString(p, "etag", etag);
+
+    jsonEnd(p);
+    finishResponse(response);
+  }
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2781]: if rc < 0, you still need to respond right? be sure that either it already did respond, or that you do it here.

File: c/datasetjson.c [hunk 25]
```diff
@@ -2501,9 +2643,120 @@ bool checkIfDatasetExists(char* dataset) {
...
+            JsonObject *jsonMemberObject = jsonAsObject(memberObject);
+            char* memName = jsonObjectGetString(jsonMemberObject,"name");
+            // memset(newMemberName, '\0', sizeof(newMemberName));
+            sprintf(newMemberName, "//'%s(%s)'", updatedTargetDataset, memName);
+            sprintf(sourceMemberName, "//'%s(%s)'", updatedSourceDataset, memName);
+
+            int reasonCode = 0;
+            int rc = createDataset(response, newMemberName, NULL, 0, &reasonCode);
+            if(rc != 0) {
+              return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2835]: slh free needed before return.

File: c/datasetjson.c [hunk 26]
```diff
@@ -2501,9 +2643,120 @@ bool checkIfDatasetExists(char* dataset) {
...
+            }
+
+            char msgBuffer[128];
+            char etag[128];
+
+            rc = readWriteToDatasetAndRespond(response, sourceMemberName, newMemberName, false, msgBuffer, etag);
+            if(rc < 0) {
+              // respondWithError(response, HTTP_STATUS_INTERNAL_SERVER_ERROR, "Failed to copy the dataset");
+              return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2844]: slh free needed before return.

**Review by 1000TurquoisePogs — CHANGES_REQUESTED** [review]
Just a few things here. 
You should check over your code to make sure finishresponse() is done in both success & failure flows, because its hard to read to see if a finishresponse() is done in a helper function.

Which makes me think, are helper functions that finishresponse() a little detrimental because it makes it harder to know if you did the right thing?

File: c/datasetjson.c (hunk 24 above)
Comment (sakshibobade21) [reply c/datasetjson.c:2781]: If rc<0, we respond in the function` readWriteToDatasetAndRespond`

File: c/datasetjson.c [hunk 27]
```diff
@@ -2806,65 +3479,130 @@ void newDatasetMember(HttpResponse* response, DatasetName* datasetName, char* ab
...
+      JsonObject * jsonObject = jsonAsObject(json);
+      returnCode = setTextUnit(TEXT_UNIT_STRING, DATASET_NAME_LEN, &datasetName.value[0], 0, DALDSNAM, &configsCount, inputTextUnit);
+      if(returnCode == 0) {
+        returnCode = setTextUnit(TEXT_UNIT_STRING, DD_NAME_LEN, ddNameBuffer, 0, DALDDNAM, &configsCount, inputTextUnit);
+      }
+      if(returnCode == 0) {
+        returnCode = setDatasetAttributesForCreation(jsonObject, &configsCount, inputTextUnit);
+      }
+    }
+  } else {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:3680]: needs slhfree

File: c/datasetjson.c [hunk 28]
```diff
@@ -2806,65 +3479,130 @@ void newDatasetMember(HttpResponse* response, DatasetName* datasetName, char* ab
...
+    while (*reasonCode==0x4100000 && ddNumber < 100000) {
+      sprintf(ddNameBuffer, "MVD%05d", ddNumber);
+      int ddconfig = 1;
+      setTextUnit(TEXT_UNIT_STRING, DD_NAME_LEN, ddNameBuffer, 0, DALDDNAM, &ddconfig, inputTextUnit);
+      returnCode = dynallocNewDataset(inputTextUnit, configsCount, reasonCode);
+      ddNumber++;
+    }
+  }
+
+  if (returnCode) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:3698]: needs slhfree

File: c/datasetjson.c [hunk 29]
```diff
@@ -2806,65 +3479,130 @@ void newDatasetMember(HttpResponse* response, DatasetName* datasetName, char* ab
...
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_WARNING,
+            "error: ds alloc dsn=\'%44.44s\' dd=\'%8.8s\', sysRC=%d, sysRSN=0x%08X\n",
+            daDatasetName.name, ddNameBuffer, returnCode, *reasonCode);
+    respondWithError(response, HTTP_STATUS_INTERNAL_SERVER_ERROR, "Unable to allocate a DD for ACB");
+    return ERROR_ALLOCATING_DATASET;
+  }
+
+  memcpy(daDDName.name, ddNameBuffer, DD_NAME_LEN);
+  daRC = dynallocUnallocDatasetByDDName(&daDDName, DYNALLOC_UNALLOC_FLAG_NONE, &returnCode, reasonCode);
+  if (daRC != RC_DYNALLOC_OK) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:3709]: needs slhfree

File: c/datasetjson.c [hunk 30]
```diff
@@ -2806,65 +3479,130 @@ void newDatasetMember(HttpResponse* response, DatasetName* datasetName, char* ab
...
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_WARNING,
+            "error: ds unalloc dsn=\'%44.44s\' dd=\'%8.8s\', rc=%d sysRC=%d, sysRSN=0x%08X\n",
+            daDatasetName.name, daDDName.name, daRC, returnCode, *reasonCode);
+    respondWithError(response, HTTP_STATUS_INTERNAL_SERVER_ERROR, "Unable to deallocate DDNAME");
+    return ERROR_DEALLOCATING_DATASET;
+  }
+  return 0;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:3718]: needs slhfree

**Review by 1000TurquoisePogs — CHANGES_REQUESTED** [review]
Need to fix memory leaks

File: c/datasetjson.c [hunk 31]
```diff
@@ -2806,65 +3479,130 @@ void newDatasetMember(HttpResponse* response, DatasetName* datasetName, char* ab
...
+  #endif
+}
+
+void createDatasetAndRespond(HttpResponse* response, char* absolutePath, int jsonMode) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:3722]: this function has a safeMalloc() with no safeFree().

**Review by 1000TurquoisePogs — CHANGES_REQUESTED** [review]

File: c/datasetjson.c [hunk 32]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  // To set attributes for target dataset
+  char datasetAttributes[300];
+  int isPDS = setAttrForDSCopyAndRespondIfError(response, datasetAttrBuffer, datasetAttributes, !isSourceMemberEmpty);
+
+  // Pasting as a dataset member [PS -> Member OR Member -> Member]
+  if(!isTargetMemberEmpty && (!isPDS || !isSourceMemberEmpty)){
+    int targetDsnExists = checkIfDatasetExistsAndRespond(response, targetDataset, false);
+    if(targetDsnExists < 0) {
+      return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2957]: return without free.
needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 33]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+    } else if(targetDsnExists == 0) {
+      respondWithJsonError(response, "Cannot paste member. Target dataset does not exist.", 400, "Bad Request");
+      return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2961]: return without free.
needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 34]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+    } else if(targetDsnExists == 1) {
+      return pasteAsDatasetMember(response, sourceDataset, targetDataset);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2964]: return without free.
needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 35]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  if(rc != 0) {
+    return;
+  }
+
+  if(isPDS == 1) {
+    // Paste the entire PDS(E) directory
+    if(isTargetMemberEmpty) {
+      pastePDSDirectory(response, datasetAttrBuffer, sourceDataset, targetDataset);
+      return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2981]: return without free.
needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 36]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+    } else {
+      respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Invalid Target. Cannot paste PDS(E) as a member");
+      return;
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2985]: return without free.
needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 37]
```diff
@@ -2225,211 +2268,839 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
 
-  HttpRequestParam *listMembersParam = getCheckedParam(request,"listMembers");
-  char *listMembersArg = (listMembersParam ? listMembersParam->stringValue : NULL);
+  DatasetName targetDsnName;
+  DatasetMemberName targetMemName;
+  extractDatasetAndMemberName(targetDataset, &targetDsnName, &targetMemName);
 
-  int datasetTypeCount = (typesArg == NULL) ? 3 : strlen(typesArg);
+  // Buffer to save attributes for target dataset
+  JsonBuffer *datasetAttrBuffer = makeJsonBuffer();
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2472]: i dont see a safeFree on this anywhere. Missed?

**Review by 1000TurquoisePogs — CHANGES_REQUESTED** [review]

File: c/datasetjson.c [hunk 38]
```diff
@@ -2336,210 +2376,850 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+    } else if(targetDsnExists == 1) {
+      safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);
+      return pasteAsDatasetMember(response, sourceDataset, targetDataset);
+    }
+  }
+
+  int reasonCode = 0;
+  int rc = createDataset(response, targetDataset, datasetAttributes, strlen(datasetAttributes), &reasonCode);
+
+  if(rc != 0) {
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c:2971]: needs safeFree((char*)datasetAttrBuffer, datasetAttrBuffer->size);

File: c/datasetjson.c [hunk 39]
```diff
@@ -2336,210 +2376,850 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
...
+
+  int datasetCount = 0;
+
+  // Buffer to save attributes for target dataset
+  JsonBuffer *datasetAttrBuffer = makeJsonBuffer();
+  jsonPrinter *jPrinter = makeBufferNativeJsonPrinter(CCSID_UTF_8, datasetAttrBuffer);
+  jsonStart(jPrinter);
+
+   // To get the attributes for target dataset
+  getDatasetMetadata(&dsnName, &memName, dataset, "true", "true", defaultDatasetTypesAllowed, "true", 0, NULL, NULL, "", NULL, jPrinter);
```
Comment (1000TurquoisePogs) [inline c/datasetjson.c]: 5th argument true=get details, 7th argument true=get members.
you dont need details or members to check if the dataset exists, its a waste of processing.

**Review by 1000TurquoisePogs — APPROVED** [review]

**Comment by 1000TurquoisePogs** [comment]
After review I find no more occurrences of memory issues or http requests missing responses, and I've seen demos of this working as intended so I'll merge it now.
